src/stepfunctions/steps/sagemaker.py

```diff
@@ -103,6 +116,16 @@ def __init__(self, state_id, estimator, job_name, data=None, hyperparameters=Non
         if isinstance(job_name, (ExecutionInput, StepInput)):
             parameters['TrainingJobName'] = job_name
 
+        if output_path is not None:
+            if isinstance(output_path, (ExecutionInput, StepInput)) or isinstance(output_path, str):
+                parameters['OutputDataConfig']['S3OutputPath'] = output_path
+
+        if data is not None and data_placeholder:
+            # Replace the 'S3Uri' key with one that supports JSONpath value.
```

Comment:
This is already done here isn't it?
https://github.com/aws/aws-step-functions-data-science-sdk-python/blob/main/src/stepfunctions/steps/states.py#L51-L77

Reply:
Nevermind. `parameters` becomes the output of [`sagemaker.workflow.airflow.training_config`](https://sagemaker.readthedocs.io/en/stable/workflows/airflow/sagemaker.workflow.airflow.html#training-config)  from L106-108 which doesn't have Placeholders.

Reply:
Yes, it is. For this case, to generate the training parameters, we need to call [sagemaker.workflow.airflow.training_config](https://sagemaker.readthedocs.io/en/stable/workflows/airflow/sagemaker.workflow.airflow.html) with _inputs=data_ which does not accept [ExecutionInput](https://aws-step-functions-data-science-sdk.readthedocs.io/en/stable/placeholders.html?highlight=executionInput#stepfunctions.inputs.ExecutionInput) or [StepInput](https://aws-step-functions-data-science-sdk.readthedocs.io/en/stable/placeholders.html?highlight=executionInput#stepfunctions.inputs.StepInput) objects for _inputs_. To bypass that, if _data_ is either an [ExecutionInput](https://aws-step-functions-data-science-sdk.readthedocs.io/en/stable/placeholders.html?highlight=executionInput#stepfunctions.inputs.ExecutionInput) or [StepInput](https://aws-step-functions-data-science-sdk.readthedocs.io/en/stable/placeholders.html?highlight=executionInput#stepfunctions.inputs.StepInput) object, we transform it to _JSONpath_ before calling [training_config()](https://sagemaker.readthedocs.io/en/stable/workflows/airflow/sagemaker.workflow.airflow.html) (which will be processed as a string). Doing that, the parameters will be generated with the JSONpath as uri, but the corresponding key won't be formatted to take JSONpath value (it will be missing the ".$" at the end).

Alternatively, we could also modify [_replace_placeholders()](https://github.com/aws/aws-step-functions-data-science-sdk-python/blob/main/src/stepfunctions/steps/states.py#L51-L65) to replace the key (and add ".$") when the value is already a JSONpath. That could be a cleaner way to do it and could be re-used as we make more parameters placeholder compatible.

Reply:
Yes, I figured that out myself after reading more of the code.

> Alternatively, we could also modify _replace_placeholders() to replace the key (and add ".$") when the value is already a JSONpath. That could be a cleaner way to do it and could be re-used as we make more parameters placeholder compatible.

If you're talking about replacing the keys regular strings which happen to be JSONPath strings, we can't do that because it's backwards incompatible. It's valid usage to pass a normal string `"$.that.is.jsonpath"` as a value for any parameter.

Reply:
+1
Agreed! Let's keep it that way and maybe add checks to make sure the accessed keys in parameters exist.